### ISSUE TITLE
FLAG Data Augmentation Approach 

### ISSUE THREAD

**Issue by reislerjul** [pyg-team/pytorch_geometric#6795]
### 🚀 The feature, motivation and pitch

Myself and a few collaborators would like to implement [FLAG](https://arxiv.org/pdf/2010.09891.pdf), an adversarial data augmentation strategy, and are looking for feedback on our architecture. 

We propose a FLAG module:
```python
class FLAG(torch.nn.Module)
```
where `FLAG` is instantiated with a number of hyperparameters, and the FLAG algorithm is run in the `forward` method. A training loop that uses FLAG could look something like this:

```python
flag = FLAG(optimizer, loss_fn, device)

for epoch in range(num_epochs):
    loss, out = flag(x, edge_index, y_true, model, train_idx, step_size, n_ascent_steps)
```
Note that calling `optimizer.zero_grad()`, `loss.backward()`, and `optimizer.step()` are handled in FLAG's `forward` method. Our code would be adapted from:
https://github.com/devnkong/FLAG/blob/main/deep_gcns_torch/examples/ogb/ogbn_arxiv/main.py#L56
https://github.com/devnkong/FLAG/blob/main/deep_gcns_torch/examples/ogb/attacks.py

We are planning to add it to the `contrib` package but have thought about putting it in `torch_geometric.utils.augmentation`. 

Any suggestions on architecture or where this should live would be helpful! 

### Alternatives

We considered extending `BaseTransform` but believe that approach wouldn't work since FLAG iteratively updates the perturbation, requires access to the model, optimizer, and loss function, and performs gradient updates. 

### Additional context

Below is the FLAG algorithm in pseudocode:
<img width="574" alt="algorithm" src="https://user-images.githubusercontent.com/17657019/221332013-168e7e5a-7662-4ab3-b660-28fb875976f8.png">

Note that for now, we are only planning to handle graphs with continuous node features. In the FLAG paper, to handle discrete node features, the authors project the features into a continuous space and then augment the hidden embeddings. However, doing this requires model architecture changes. If there are any suggestions on extending the architecture for graphs with discrete node features, please let us know!

**Comment by rusty1s** [comment]
Thanks for the issue. I think this should be indeed be a model rather than a transform. Since the training logic seems complex, I think it makes sense to do this as part of `forward`.
